### ISSUE TITLE
Secret update caused massive SecretTrigger

### ISSUE THREAD

**Issue by shonecyx** [istio/istio#51899]
**Describe the feature request**

Currently [NewCredentialsController](https://github.com/istio/istio/blob/master/pilot/pkg/credentials/kube/secrets.go#L79) watches secrets in the mesh and doesn't honor the discoverySelector:
```
        fieldSelector := fields.AndSelectors(
		fields.OneTermNotEqualSelector("type", "helm.sh/release.v1"),
		fields.OneTermNotEqualSelector("type", string(v1.SecretTypeServiceAccountToken))).String()
	secrets := kclient.NewFiltered[*v1.Secret](kc, kclient.Filter{
		FieldSelector: fieldSelector,
```
In our mesh, there is churn for the `opaque` secret, not very high but around 1.7k per two minutes:
![Screenshot 2024-07-04 at 18 59 07](https://github.com/istio/istio/assets/12857255/c29a3844-aa50-4c09-815b-8236d3618d48)

However that rate of churn caused millions of secret push triggers:
![Screenshot 2024-07-04 at 19 00 40](https://github.com/istio/istio/assets/12857255/d0c0abcd-2008-4741-98a5-71928589081c)

There are around 4.5k sidecar in the mesh and 12 istiod pods, so each istiod has around 400 XDS connections.

![Screenshot 2024-07-04 at 19 02 33](https://github.com/istio/istio/assets/12857255/740daecf-cf4b-4a9d-a559-e92452981166)

The root cause of the millions of secret push trigger is that, for each secret update, it's a kind of configure update and will trigger each[ istiod full push](https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L600C1-L614C1) to all xds clients:
```
// Send a signal to all connections, with a push event.
func (s *DiscoveryServer) StartPush(req *model.PushRequest) {
	// Push config changes, iterating over connected envoys.
	if log.DebugEnabled() {
		currentlyPending := s.pushQueue.Pending()
		if currentlyPending != 0 {
			log.Debugf("Starting new push while %v were still pending", currentlyPending)
		}
	}
	req.Start = time.Now()
	for _, p := range s.AllClients() {
		s.pushQueue.Enqueue(p, req)
	}
}

```

So the 1.7k churn will  increase 400 times due to each secret churn will cause push trigger to all proxies connected to the istiod and result to millions of secret trigger.

**Proposed solutions**
1.  Add an option to exclude the Opaque secret since in our case we don't need istiod to watch Opaque secret.
```
fields.OneTermNotEqualSelector("type", string(v1.SecretTypeOpaque)),
```
2. When istiod sends push to all clients, it can check if the trigger reason is secret and will enque only if the secret namespace matches proxy namespace(ideally secret should not be cross namespace referred), some draft code like below:
```
        for _, p := range s.AllClients() {
		if len(req.Reason) == 1 && req.Reason[0] == model.SecretTrigger {
			for c := range req.ConfigsUpdated {
				if len(req.ConfigsUpdated) == 1 && c.Namespace != p.proxy.Metadata.Namespace {
					contiue;
				}
			}
		}
		s.pushQueue.Enqueue(p, req)
	}
``` 

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[ x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

**Comment by howardjohn** [comment]
is the queue specifically a bottle neck? most of our filtering happens after the queue

**Comment by shonecyx** [comment]
> is the queue specifically a bottle neck? most of our filtering happens after the queue

@howardjohn I think the bottle neck is not the queue since the the proxy queue time is good. But for this case, the problem is that secret churn will cause a lot of unnecessary pushes and there is no filtering after the queue.  
We can either filter before the enqueue to add another `OneTermNotEqualSelector` for opaque secret, or after enqueue do some filtering based on the secret namespace and proxy namespace.

**Comment by howardjohn** [comment]
A secret update should be an incremental push. You said its a full push. Can you expand more?

**Comment by shonecyx** [comment]
@howardjohn 
You are right, although the secret update is not a [full push](https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/server.go#L514-L521), it will cause push to all connections:
```
creds.AddSecretHandler(func(name string, namespace string) {
	s.XDSServer.ConfigUpdate(&model.PushRequest{
		Full:           false,
		ConfigsUpdated: sets.New(model.ConfigKey{Kind: kind.Secret, Name: name, Namespace: namespace}),

		Reason: model.NewReasonStats(model.SecretTrigger),
	})
})
```
And it will trigger [AdsPushAll](https://github.com/istio/istio/blob/master/pilot/pkg/xds/discovery.go#L252-L258)
```
if !req.Full {
  req.Push = s.globalPushContext()
  s.dropCacheForRequest(req)
  s.AdsPushAll(req)
  return
}
```
And [AdsPushAll](https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L560C1-L577C2) will just log `!req.Full` then startPush:
```
func (s *DiscoveryServer) AdsPushAll(req *model.PushRequest) {
	if !req.Full {
		log.Infof("XDS: Incremental Pushing ConnectedEndpoints:%d Version:%s",
			s.adsClientCount(), req.Push.PushVersion)
	} else {
		totalService := len(req.Push.GetAllServices())
		log.Infof("XDS: Pushing Services:%d ConnectedEndpoints:%d Version:%s",
			totalService, s.adsClientCount(), req.Push.PushVersion)
		monServices.Record(float64(totalService))

		// Make sure the ConfigsUpdated map exists
		if req.ConfigsUpdated == nil {
			req.ConfigsUpdated = make(sets.Set[model.ConfigKey])
		}
	}

	s.StartPush(req)
}
```
Which will trigger `push event` to all [connections](https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L580-L592):
```
// Send a signal to all connections, with a push event.
func (s *DiscoveryServer) StartPush(req *model.PushRequest) {
	// Push config changes, iterating over connected envoys.
	if log.DebugEnabled() {
		currentlyPending := s.pushQueue.Pending()
		if currentlyPending != 0 {
			log.Debugf("Starting new push while %v were still pending", currentlyPending)
		}
	}
	req.Start = time.Now()
	for _, p := range s.AllClients() {
		s.pushQueue.Enqueue(p, req)
	}
}
```
Normally there won't be so many secret churn in the cluster, but we are heavliy using secret which caused this issue. 

Is this expected that any secret update in the mesh will cause push to all proxies?

**Comment by howardjohn** [comment]
Yes, a incremental push is supposed to be cheap. Are you seeing perf issues? do you have a flamegraph if so

**Comment by shonecyx** [comment]
@howardjohn  FYI the flamegraph I just captured:
![Screenshot 2024-07-09 at 11 39 15](https://github.com/istio/istio/assets/12857255/ac2f0dc5-0212-4525-8d37-fcb268315046)
![Screenshot 2024-07-09 at 11 37 26](https://github.com/istio/istio/assets/12857255/59046128-bfec-4569-b433-c834bbcab1cc)

**Comment by howardjohn** [comment]
so looks like no time spent caused by secrets? unless I misunderstood

**Comment by shonecyx** [comment]
@howardjohn  Yes, it's more like a metrics issue rather than a performance issue, although in theory it will impact istiod push performance since more events in the pushQueue need to handle, it's not obvious in the flamegraph.

And it actually got filtered in the [pushConnection](https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L481-L484)  and the sdsservice.ProxyNeedsPush will actually skip the request and this logic got refactored in [here](https://github.com/istio/istio/pull/50828/files#diff-37776d41e7fb1a802dd06dea3af99b63270d8ec780daf581e8e2748ff6c22b25L83): 
```
if !s.ProxyNeedsPush(con.proxy, pushRequest) {
  log.Debugf("Skipping push to %v, no updates required", con.ID())
  return nil
}
```
Come back to the SecretTrigger metrics, do you think it makes sense to make it on par with the proxies number that it real pushed or just do some merge based on the secret name/namespace.

**Comment by howardjohn** [comment]
The metric is the number of push _triggers_ not the number of pushes, so I don't think so. There is already another metric for the number of pushes?

**Comment by howardjohn** [comment]
It is not accurate that it is only for errors. https://github.com/istio/istio/blob/d829a4ee0a193997497c5455c7f7cf0518a6e2c6/pilot/pkg/xds/delta.go#L504

**Comment by howardjohn** [comment]
Is there anything pending here? This seems working as intended

**Comment by shonecyx** [comment]
> Is there anything pending here? This seems working as intended

@howardjohn  Shall we do some enhancement to reduce the unnecessary secretPush triggers? Although it's just enque and deque without real push, the metrics here is a little bit confusing.

**Comment by howardjohn** [comment]
I do not think so, if you want a metric indicating the number of pushes there is already one. I think you may just be looking at the wrong metric for your usage

**Comment by shonecyx** [comment]
Thanks @howardjohn, the secret push trigger metrics caused some confusing.
